benchmarks/performance-poetry/complex-poetry/src/main/java/com/squareup/benchmarks/performance/complex/poetry/PerformancePoetryComponent.kt

```diff
@@ -37,9 +45,121 @@ class PerformancePoetryComponent : ViewModel() {
     browserIsLoading.combine(poemIsLoading) { one, two -> one || two }
   )
 
+  data class RenderStats(
+    var nodesRenderedTheSame: Int = 0,
+    var nodesRenderedDifferent: Int = 0
+  )
+
+  private val performanceTraceInterceptor = object : WorkflowInterceptor {
+    var totalRenderPasses = 1
+    var totalFresh = 0
+    var totalStale = 0
+    var rootSessionId: Long = -1
+    val sessionRenderStats = mutableMapOf<Long, RenderStats>()
+    val nodeStates: MutableMap<Long, String> = mutableMapOf()
+
+    @SuppressLint("BinaryOperationInTimber")
+    override fun <P, S, O, R> onRender(
+      renderProps: P,
+      renderState: S,
+      context: BaseRenderContext<P, S, O>,
+      proceed: (P, S, RenderContextInterceptor<P, S, O>?) -> R,
+      session: WorkflowSession
+    ): R {
+      val root = session.parent == null
+      val tracedId = NODES_TO_TRACE.firstOrNull { it == session.identifier } != null
+      if (root) {
+        // Setup stats for render pass tracking.
+        rootSessionId = session.sessionId
+        sessionRenderStats[rootSessionId] = RenderStats()
+        // Cumulative Section that has total render pass time.
+        Trace.beginSection("RenderPassTotal")
+        Trace.beginSection("RenderPass-$totalRenderPasses")
```

Comment:
Are you certain these will always close in the proper order and there won't be any additional sections that intersect (e.g. the ending doesn't happen from a posted main thread message)?

If not you could use async sections.

Reply:
In my testing they were nested properly by the nature of the WOrkflowInterceptor - but do you have any reason to doubt that based on your work with teh Workflow Interceptor?